lib/chef/resource/cpan_module.rb

```diff
@@ -0,0 +1,143 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require "chef/resource"
+
+class Chef
+  class Resource
+    class CpanModule < Chef::Resource
+      resource_name :cpan_module
+      provides :cpan_module
+
+      description "Use the cpan_module resource to install or uninstall perl CPAN modules."
+      introduced "14.0"
+
+      property :module_name,
+               String,
+               description: "The name of the module if it's different than the name of the resource.",
+               name_property: true
+
+      property :force,
+               [TrueClass, FalseClass],
+               description: "To force the install within cpanm.",
+               default: false
+
+      property :test,
+               [TrueClass, FalseClass],
+               description: "To do a test install.",
+               default: false
+
+      property :version,
+               String,
+               description: "Any version string cpanm would find acceptable."
+
+      property :cwd,
+               String,
+               description: "A path to change into before running cpanm."
+
+      property :cpanm_binary,
+               String,
+               description: "The path of the cpanm binary.",
+               default: "cpanm"
+
+      action :install do
+        description "Install the module."
+
+        declare_resource(:execute, "CPAN :install #{new_resource.module_name}") do
+          cwd current_working_dir
+          command cpanm_install_cmd
+          environment "HOME" => current_working_dir, "PATH" => "/usr/local/bin:/usr/bin:/bin"
+          not_if { module_exists_new_enough }
+        end
+      end
+
+      action :uninstall do
+        description "Uninstall the module."
+
+        declare_resource(:execute, "CPAN :uninstall #{new_resource.module_name}") do
+          cwd current_working_dir
+          command cpanm_uninstall_cmd
+          only_if { module_exists? }
+        end
+      end
+
+      action_class do
+        def module_exists_new_enough
+          existing_version = parse_cpan_version
+          return false if existing_version.empty? # mod doesn't exist
+          return true if new_resource.version.nil? # mod exists and version is unimportant
+          @comparator, @pending_version = new_resource.version.split(" ", 2)
+          @current_vers = Gem::Version.new(existing_version)
+          @pending_vers = Gem::Version.new(@pending_version)
+          @current_vers.method(@comparator).call(@pending_vers)
```

Comment:
This should use`Gem::Requirement.new().satisfied_by?` API instead of what its doing here.